hyperspy/learn/onmf.py

```diff
@@ -0,0 +1,241 @@
+        if store_r:
+            self.R = []
+        else:
+            self.R = None
+
+    def _setup(self, X, normalize=False):
+        self.h, self.r = None, None
+        if isinstance(X, np.ndarray):
+            n, m = X.shape
+            if normalize:
+                self.X_min = X.min()
+                self.X_max = X.max()
+                self.normalize = normalize
+                # actually scale the data to be between 0 and 1, not just close
+                # to it..
+                X = _normalize(X, ar_min=self.X_min, ar_max=self.X_max)
+                # X = (X - self.X_min) / (self.X_max - self.X_min)
+            avg = np.sqrt(X.mean() / m)
+        else:
+            if normalize:
+                _logger.warning("Normalization with an iterator is not"
+                                " possible, option ignored.")
+            x = next(X)
+            m = len(x)
+            avg = np.sqrt(x.mean() / m)
+            X = chain([x], X)
+
+        self.nfeatures = m
+
+        self.W = np.abs(avg * halfnorm.rvs(size=(self.nfeatures, self.rank)) /
+                        np.sqrt(self.rank))
+
+        self.A = np.zeros((self.rank, self.rank))
+        self.B = np.zeros((self.nfeatures, self.rank))
+        return X
+
+    def fit(self, X, batch_size=None):
+        if self.nfeatures is None:
+            X = self._setup(X)
+
+        num = None
+        prod = np.outer
+        if batch_size is not None:
+            prod = np.dot
+            length = X.shape[0]
+            num = max(length // batch_size, 1)
+            X = np.array_split(X, num, axis=0)
+        if isinstance(X, np.ndarray):
+            num = X.shape[0]
+            X = iter(X)
+        r, h = self.r, self.h
+        for v in progressbar(X, leave=False, total=num, disable=num == 1):
+            h, r = _solveproj(v, self.W, self.lambda1, self.kappa, r=r, h=h)
+            self.H.append(h)
+            if self.R is not None:
+                self.R.append(r)
+
+            self.A += prod(h, h.T)
+            self.B += prod((v.T - r), h.T)
+            self._solve_W()
+        self.r = r
+        self.h = h
+
+    def _solve_W(self):
+        eta = self.kappa / np.linalg.norm(self.A, 'fro')
+        if self.robust:
+            # exactly as in the paper
+            n = 0
+            lasttwo = np.zeros(2)
+            while n <= 2 or (np.abs(
+                (lasttwo[1] - lasttwo[0]) / lasttwo[0]) > 1e-5 and n < 1e9):
+                self.W -= eta * (np.dot(self.W, self.A) - self.B)
+                self.W = _project(self.W)
+                n += 1
+                lasttwo[0] = lasttwo[1]
+                lasttwo[1] = 0.5 * np.trace(self.W.T.dot(self.W).dot(self.A)) - \
+                        np.trace(self.W.T.dot(self.B))
+        else:
+            # Tom's approach
+            self.W -= eta * (np.dot(self.W, self.A) - self.B)
+            np.maximum(self.W, 0.0, out=self.W)
+            self.W /= max(np.linalg.norm(self.W, 'fro'), 1.0)
```

Comment:
I've not tested this, but am I right in thinking that:

- My approach takes "robust" to mean an l2-norm cost function that can deal with sparse corruptions/outliers (i.e. the same meaning as RPCA) - see https://arxiv.org/abs/1604.02634
- Your interpretation of `self.robust` deals with generalized cost functions - see https://arxiv.org/abs/1608.00075

Is that right? If so, how do the results of the two approaches differen, and when would you use one over the other? Is one much faster than the other for example?